### DISCUSSION TITLE
SSR - traverse the tree and prefetch all data

### DISCUSSION THREAD

**Discussion by zsolt-dev** [TanStack/query#3270] [Ideas] [answered]
Hi,

I am new to react-query, so not sure I understand it correctly.

This is the SSR example from your documentation:

```
 import { dehydrate, Hydrate, QueryClient, QueryClientProvider } from 'react-query';
2 
3 function handleRequest (req, res) {
4   const queryClient = new QueryClient()
5   await queryClient.prefetchQuery('key', fn)
6   const dehydratedState = dehydrate(queryClient)
7 
8   const html = ReactDOM.renderToString(
9     <QueryClientProvider client={queryClient}>
10       <Hydrate state={dehydratedState}>
11         <App />
12       </Hydrate>
13     </QueryClientProvider>
14   )
15 
16   res.send(`
17     <html>
18       <body>
19         <div id="root">${html}</div>
20         <script>
21           window.__REACT_QUERY_STATE__ = ${JSON.stringify(dehydratedState)};
22         </script>
23       </body>
24     </html>
25   `)
26 }
```

As you can see, the current SSR implementation requires to manually prefetch all the data:

**`await queryClient.prefetchQuery('key', fn)`**

It would be nice to have it like apollo-client, that walks down the entire tree and executes every required query it encounters:

```
import { getDataFromTree } from "@apollo/client/react/ssr";

// This function walks down the entire tree and executes every required query it encounters (including nested queries). 
getDataFromTree(App).then((content) => {
  // Extract the entirety of the Apollo Client cache's current state
  const initialState = client.extract();

  // Add both the page content and the cache state to a top-level component
  const html = <Html content={content} state={initialState} />;

  // Render the component to static markup and return it
  res.status(200);
  res.send(`<!doctype html>\n${ReactDOM.renderToStaticMarkup(html)}`);
  res.end();
});
```

Is there a plan to implement a function like **getDataFromTree()** in react-query ?

Thank you

**Comment by TkDodo** [comment]
I have no idea how this is implemented in apollo, but it's a very cool concept. If you know how to so it, let's talk about a contribution :) I'm guessing since this is server side code, it wouldn't even affect client bundle size, right?

@Ephem fyi

**Reply by zsolt-dev** [reply]
Yes bundle size will not be affected.

Here is how its implemented in Apollo:
https://github.com/apollographql/apollo-client/blob/main/src/react/ssr/getDataFromTree.ts

I do not know react-query at all so I would not know how to implement it.

This would essentially solve the SSR problem everyone is asking about in the discussion. 

Thank you

**Reply by TkDodo** [reply]
Right, I don't really see how apollo is doing it. It looks like it just renders the static markup, but I don't see how it knows which queries to execute. If we do that with RQ, I think we would have our cache prefilled with optimistic results - mostly empty `loading` states. So we would then need to pick those up and execute the queryFns.

It seems doable, however, I have one concern:

fetching on the server doesn't have to be the same as fetching on the client. I would argue that a lot of the time, it shouldn't be. We don't really communicate this in the docs, as our example is:

```
 export async function getStaticProps() {
   const queryClient = new QueryClient()
 
   await queryClient.prefetchQuery('posts', getPosts)
 
   return {
     props: {
       dehydratedState: dehydrate(queryClient),
     },
   }
 }
 
 function Posts() {
   const { data } = useQuery('posts', getPosts)
 }
```

where we use the same `queryFn` (`getPosts`) on the server and on the client. However, on the server, it could be a different thing, like:

```
await queryClient.prefetchQuery('posts', getPostsDirectlyFromDB)
```

all that matters is that the result shape is the same. On the frontend, we have to go and point to our api endpoint that then goes to the database. Maybe it's irrelevant, but that tree-walking mechanism would basically be forced to execute the function from the client also on the server. Maybe that's irrelevant, but maybe it isn't ...

**Reply by zsolt-dev** [reply]
Good points.

Everything was already sorted out by the apollo team:
[Avoiding the network for local queries](https://www.apollographql.com/docs/react/performance/server-side-rendering/#avoiding-the-network-for-local-queries)

Also skipping queries that should be only executed on the client etc...

**Reply by TkDodo** [reply]
> Everything was already sorted out by the apollo team:

yes, for graphQL specifically :) while `ssr: false` is something we could also adopt, I don't see how we could solve the first example where we want completely different queryFns. But maybe it's just a limitation we could accept

**Comment by Ephem** [comment, marked as answer]
I think this is already possible today using React Query with Suspense + [react-ssr-prepass](https://github.com/FormidableLabs/react-ssr-prepass), which is a more performant implementation of a prepass render step than Apollo's one (at least last I checked). 🎉

When React 18 is released, while it wont have official support for Suspense data fetching, I think it might work out of the box with React Query as long as you are not streaming the response (or at least, as long as you don't hydrate the markup until you have the full response). If it does not work out of the box, I think that's where we should spend our effort instead of adding a prepass step, which is usually terrible for SSR performance.

Also, even with Suspenseful SSR, for the best performance you will still want to start fetching those queries as soon as possible to avoid request waterfalls, so those prefetches will still be the best practice going forward, it's just that you can _also_ start rendering earlier on the server. 😃

**Reply by zsolt-dev** [reply]
Thank you @Ephem.

I confirm it works.

**Reply by TkDodo** [reply]
if that is the way to go - can we put that in the docs on the SSR page? I think it would be great to let people know that this option exists!

**Reply by zsolt-dev** [reply]
Sure, good idea. Leaving some more info: 

There is one small problem with it. React SSR (renderToString....) does not support ErrorBoundary.

So in case an Error thrown from react-query, it is caught not by react itself but by Expressjs, so the whole page return error 500.

Hopefully react 18 will support ErrorBoundaries with SSR.

By the way, we use it with preact and with preact-ssr-prepass and works the same as with standard react.

**Reply by Ephem** [reply]
Nice! Yeah, this is a known problem, and I don't think they intend to support ErrorBoundaries on the server because of fundamental problems with that when it comes to streaming. There was great guide posted about this in the React 18 working group recently, and it has been mentioned in other posts there as well: [Handling Root Errors in SSR Streaming Rendering](https://github.com/reactwg/react-18/discussions/122)

**Reply by TkDodo** [reply]
@zsolt-dev feel free to open a PR to the docs please.

**Reply by Uttu316** [reply]
Is there any tutorial that can help to use [react-ssr-prepass](https://github.com/FormidableLabs/react-ssr-prepass) with React query for multiple api calls?

**Reply by TkDodo** [reply]
I was hoping that @zsolt-dev could contribute something to the docs or examples :)

**Reply by zsolt-dev** [reply]
@TkDodo Will prepare it during the weekend :)

**Reply by PrescottJRynewicz** [reply]
Did we ever get around to this @zsolt-dev?

**Reply by zsolt-dev** [reply]
@PrescottJRynewicz 

Submitted a [PR](https://github.com/TanStack/query/pull/4182) just now :)

**Comment by KATT** [comment]
If someone wants a reference implementation, I have one in tRPC [here](https://github.com/trpc/trpc/blob/main/packages/next/src/withTRPC.tsx) which works alright (and isn't using suspense. It's heavily based on urql's prepassing but isn't using suspense either.

**Comment by dosco** [comment]
This is pretty simple to implement and does not have to do with react-query or any additional package just use `React.renderTo...` functions.

```
await new Promise((onAllReady, onError) => renderToPipeableStream(el, { onAllReady, onError }));
```

```
const queryClient = new QueryClient();
const el(state) = <App queryClient={queryClient} dehydratedState={state} />;

await new Promise((onAllReady, onError) => renderToPipeableStream(el(), { onAllReady, onError }));
const dehydratedState = dehydrate(queryClient);
```

And the next time you call any of the `React.renderTo...` functions to do the real render then just pass back in the state by calling `el(state)` so that data fetch does not happen a second time. Any fetches trigged via a useEffect will be done client side and rest will be done server side

**Reply by TkDodo** [reply]
> Any fetches trigged via a useEffect will be done client side and rest will be done server side

not sure I understand that. basically, `useQuery` uses `useEffect` under the hood - so how would this do anything on the server then?

**Reply by dosco** [reply]
My bad. `useEffect` by itself does not enforce client side it will run server side as well. I use `enabled: typeof window !== "undefined"` on the `useQuery` to ensure it runs only client side.

**Reply by Ephem** [reply]
`useEffect` definitely does not run on the server (even with the new APIs), so unless you are using Suspense, this wouldn’t work. Also, if you _are_ using Suspense and the new APIs that support it, why do a prepass at all? You can just wait for the `renderToPipeableStream` and you’ll have your markup in the first pass. 😄

**Reply by dosco** [reply]
If I recall correctly the reason was because the dehydration data is set in the `<script>` tag for the client to consume using the `bootstrapScriptContent:`  option on the `ReactDOM.renderToPipeableStream` function itself so if the data it not available before `renderToPipeableStream` is called then you get empty values `<script>window.__REACT_QUERY_STATE__={"mutations":[],"queries":[]}</script>`. And yes suspense needs to be enabled.
